types/board.d.ts

```diff
@@ -0,0 +1,117 @@
+import { Set, Map, List, Record } from 'immutable';
+
+type Color = string | null;
+
+export declare const BLACK: Color;
+export declare const WHITE: Color;
+export declare const EMPTY: Color;
+
+export declare class Board extends Record(
+  { dimensions: 19, moves: Map<Coordinate, Color>() },
+  'Board'
+) {
+  constructor(dimensions: number, ...moves: (Coordinate | Color)[]);
+}
```

Comment:
There have a problem.
Because `...moves: (Coordinate | Color)[]`, for example:
`new Board(5, new Coordinate(2, 2), BLACK, BLACK);`
`new Board(5, new Coordinate(2, 2), BLACK, new Coordinate(3, 3));`
`new Board(5, new Coordinate(2, 2), new Coordinate(3, 3), BLACK);`
will pass the type checking.
I don't how to express `Array [Coordinate, Color, Coordinate, Color, ...]` in Typescript.

Reply:
Hm.  Maybe having them alternate like this wasn't a great idea.

I was following Immutable's `Map.of()` signature.

https://github.com/duckpunch/godash/blob/master/src/board.js#L24

Looks like the only failure you'll get is at [runtime](https://github.com/immutable-js/immutable-js/blob/59c291a/src/Map.js#L57) if there isn't an even number of arguments.

[According to SO](https://stackoverflow.com/questions/61155425/how-to-define-array-with-alternating-types-in-typescript), it seems that you can't get Typescript to recognize this.

I wouldn't be opposed to major-revving the version to change the API so that it's cleaner and can be described by Typescript better.  It would be a good opportunity to do another pass at the API and upgrade some stuff since I haven't touched this in awhile.

Do you have thoughts?

Reply:
In the meantime, what do you think about adding runtime checking of the parameters and keeping what you have here?  Not perfect, but it'll at least let this get merged.

Reply:
I think add runtime checking of the parameters first is great, and in the next major version you can consider rewrite APIs cleaner and can be described by Typescript better, and maybe add more APIs and make this package move convenient.

Reply:
If I wamt to change APIs, I will pack `Coordinate` and `Color` to a new type named `Move`, can considered:
- a tuple type `[Coordinate, Color]` (is array in Javascript but with fixed length and specified type)
- a object type `{coordinate: Coordinate, color: Color}`
- `Record` from immutable
and many APIs have both `coordinate` and `color` arguments can replaces to `Move`.

Reply:
To change APIs, if I were you, I may pack `Coordinate` and `Color` to a new type named `Move`, and consider it is based on:
- a tuple type (value of this type is an array in JavaScript but with fixed length and specified type)
	```ts
	type Move = [Coordinate, Color]
	const move: Move = [new godash.Coordinate(1, 0), godash.BLACK];
	```
- a interface (value of this type is an array in JavaScript but must have required properties, I recommend this)
	```ts
	interface Move {
	  coordinate: Coordinate;
	  color: Color;
	}
	const move: Move = {
	  coordinate: new godash.Coordinate(1, 0),
	  color: godash.BLACK
	};
	```
- `Record` from immutable

and many APIs have both `coordinate` and `color` arguments can be replaced to `Move` type.

And, I may consider make APIs more OOP, e.g. change
```js
let board = new godash.Board();
const coord = new godash.Coordinate(1, 0);
const color = godash.BLACK;
board = godash.addMove(board, coord, color);
```
to
```js
const board = new godash.Board();
// if type Move is based on tuple.
const move = [new godash.Coordinate(1, 0), godash.BLACK];
board.addMove(move);
```

Reply:
Seems reasonable.

Added #41 for the v1 runtime check and started a thread for v2 in #40.